cf-testd/cf-testd.c

```diff
@@ -0,0 +1,492 @@
+            break;
+        case 'f':
+            config->file = xstrdup(optarg);
+            break;
+        case 'h':
+            CFTestD_Help();
+            exit(EXIT_SUCCESS);
+        case 'I':
+            LogSetGlobalLevel(LOG_LEVEL_INFO);
+            break;
+        case 'l':
+            LoggingEnableTimestamps(true);
+            break;
+        case 'v':
+            LogSetGlobalLevel(LOG_LEVEL_VERBOSE);
+            break;
+        case 'V':
+            {
+                Writer *w = FileWriter(stdout);
+                GenericAgentWriteVersion(w);
+                FileWriterDetach(w);
+            }
+            exit(EXIT_SUCCESS);
+        default:
+            CFTestD_Help();
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    if (optind < argc) // More unparsed arguments left after getopt_long
+    {
+        // Enumerate and print all invalid arguments:
+        Log(LOG_LEVEL_ERR, "Invalid command line arguments:");
+
+        int start = optind;
+        int stop = argc;
+        int total = stop - start;
+        for (int i = 0; i<total; ++i)
+        {
+            Log(LOG_LEVEL_ERR, "[%d/%d]: %s\n", i+1, total, argv[start + i]);
+        }
+    }
+
+    return config;
+}
+
+bool CFTestD_TLSSessionEstablish(ServerConnectionState *conn)
+{
+    if (conn->conn_info->status == CONNECTIONINFO_STATUS_ESTABLISHED)
+    {
+        return 1;
+    }
+
+    bool established = BasicServerTLSSessionEstablish(conn);
+    if (!established)
+    {
+        return false;
+    }
+
+    Log(LOG_LEVEL_VERBOSE, "TLS session established, checking trust...");
+
+    /* Send/Receive "CFE_v%d" version string, agree on version, receive
+       identity (username) of peer. */
+    char username[sizeof(conn->username)] = "";
+    bool id_success = ServerIdentificationDialog(conn->conn_info,
+                                                 username, sizeof(username));
+    if (!id_success)
+    {
+        return false;
+    }
+
+    /* No CAUTH, SAUTH in non-classic protocol. */
+    conn->user_data_set = 1;
+    conn->rsa_auth = 1;
+
+    ServerSendWelcome(conn);
+    return true;
+}
+
+int CFTestD_GetServerQuery(
+    ServerConnectionState *conn, char *recvbuffer, int encrypt)
+{
+    char query[CF_BUFSIZE];
+
+    query[0] = '\0';
+    sscanf(recvbuffer, "QUERY %255[^\n]", query);
+
+    if (strlen(query) == 0)
+    {
+        return false;
+    }
+
+    return CFTestD_ReturnQueryData(conn, query, encrypt);
+}
+
+static bool CFTestD_ProtocolError(
+    ServerConnectionState *conn, const char *recvbuffer, char *sendbuffer)
+{
+    strcpy(sendbuffer, "BAD: Request denied");
+    SendTransaction(conn->conn_info, sendbuffer, 0, CF_DONE);
+    Log(LOG_LEVEL_INFO,
+        "Closing connection due to illegal request: %s", recvbuffer);
+    return false;
+}
+
+static bool CFTestD_BusyLoop(ServerConnectionState *conn)
+{
+    char recvbuffer[CF_BUFSIZE + CF_BUFEXT] = { 0 };
+    char sendbuffer[CF_BUFSIZE - CF_INBAND_OFFSET] = { 0 };
```

Comment:
Other empty string initializers are ``""``, maybe these should use the same syntax?

Reply:
Agree, fixed.